aspnetcore/testing/razor-pages-testing/sample/tests/RazorPagesTestingSample.Tests/Utilities/Utilities.cs

```diff
@@ -0,0 +1,83 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.DependencyInjection;
+using RazorPagesTestingSample.Data;
+
+namespace RazorPagesTestingSample.Tests
+{
+    public static class Utilities
+    {
+        #region snippet1
+        public class MessageComparer : IEqualityComparer<Message>
+        {
+            public bool Equals(Message firstMessage, Message secondMessage)
+            {
+                return firstMessage.Id == secondMessage.Id && 
+                    string.Equals(firstMessage.Text, secondMessage.Text, StringComparison.Ordinal);
+            }
+
+            public int GetHashCode(Message message)
+            {
+                if (Object.ReferenceEquals(message, null)) return 0;
+                int hashMessageId = message.Id.GetHashCode();
+                int hashMessageText = message.Text == null ? 0 : message.Text.GetHashCode();
+                return hashMessageId ^ hashMessageText;
+            }
+        }
+        #endregion
+
+        #region snippet2
+        public static DbContextOptions<AppDbContext> TestingDbContextOptions()
+        {
+            // Create a new service provider to create a new in-memory database.
+            var serviceProvider = new ServiceCollection()
+                .AddEntityFrameworkInMemoryDatabase()
+                .BuildServiceProvider();
+
+            // Create a new options instance using an in-memory database and 
+            // IServiceProvider that the context should resolve all of its 
+            // services from.
+            var builder = new DbContextOptionsBuilder<AppDbContext>()
+                .UseInMemoryDatabase("InMemoryDb")
+                .UseInternalServiceProvider(serviceProvider);
+
+            return builder.Options;
+        }
+        #endregion
+
+        #region snippet3
+        public static async Task<FormUrlEncodedContent> GetRequestContentAsync(
```

Comment:
Could we possibly use HtmlAgilityPack (or another Html parser) here? Parsing using regexes is kinda gross. Alternatively a comment that says, this is a sample, but we recommend using a parser here to do the right thing.

Reply:
I'm totally cool with HtmlAgilityPack ... I've used it many times. @Rick-Anderson Do you want to go with HtmlAgilityPack or Regex here?

Reply:
Like I said, as long as we guide our users to using the right thing (perhaps a comment that suggest looking for packages on nuget.org), I'm fine with what we have here.